### PR TITLE
Plot window to have better control over zoom state

### LINKED ISSUE #24476: Workbench: Plot window to have better control over zoom state
The current figure window contains a control only for returning to the top level of the zoom stack but to many users it is not obvious that this is what it does. There is also appetite for having the ability to walk up and down the zoom stack.

This issue is to assess the figure window and decide how best to provide these types of control. Some ideas:
 - Use a middle-mouse click to return to outermost zoom level
 - Add the basic forward/backward arrows from a standard matplotlib window to our figure window

### PR DESCRIPTION
**Pull request by ghost**
**Description of work.**
Several new behaviours have been added to the plot window to give users better control over the zoom state of their plots:

- The home button now always centres the contents of the figure. It previously returned the user to the top of the zoom stack, which didn't always have the contents of the figure centred.
- Forward and back buttons to move through the zoom stack.
- Scrolling now zooms in/out on the figure around the mouse position (i.e. the coordinate of the mouse remains the same when zooming). I'm interested in any feedback on this behaviour. I think this is better, and gives more control, than simply zooming around the centre of the plot, but you may disagree.
- Clicking and dragging the middle mouse button pans the figure. This saves users a click as they don't have to enter pan mode. I've left pan mode on the toolbar as a) it offers a nice "zoom in direction" mode on right-click and b) users are used to it.
- Middle-click in zoom mode centres the plot's contents (calls same function as the home button).

**To test:**
- Create a figure, you can use the below script to create one single-axes figure and one with multiple axes:
```python
import scipy.stats as stats
import math

plt.close('all')

np.random.seed(101)
mu = 0
variance = 1
sigma = math.sqrt(variance)
x = np.linspace(mu - 3*sigma, mu + 3*sigma, 100)
y = stats.norm.pdf(x, mu, sigma) + np.random.randn(100)*0.01
err = 0.01*np.ones(100)

ws = CreateWorkspace(x, y, err, OutputWorkspace='ws')
fig, ax = plt.subplots(1, 1, subplot_kw={'projection': 'mantid'})
ax.plot(ws, specNum=1)
ax.legend()
plt.show()

fig = plt.figure(figsize=[9.0, 4.7766])
for i in range(4):
    ax = fig.add_subplot(2, 2, i+1, projection='mantid')
    ax.set_title("Title {}".format(i))
    x = np.linspace(0, 10, 100)
    y = (i+1)*np.sin(x + i)
    e = np.sqrt(np.abs(y))
    ws = CreateWorkspace(x, y, DataE=e, OutputWorkspace="ws"+str(i))
    if i % 2 == 1:
        ax.errorbar(ws, label=str(i), errorevery=2, elinewidth=0.6, specNum=1)
    else:
        ax.plot(ws, label=str(i), specNum=1)
        ax.set_xlabel('x')
        ax.set_ylabel('y')
    ax.legend()
plt.show()
```

#### Home button
- Change the zoom state of the figure. Pressing the home button should centre the contents of that figure.
- Try adding new lines and adding interactive markers to the figure. The home button should work as expected.
- The home button should act on all axes if a plot has more than one.

#### Forward/backward buttons
- Zoom in/out on the figure. Use the forward and backwards buttons to move through the zoom stack. Going as far back as you can should return you to the original state.

#### Scroll zoom
- Scrolling with the mouse over a set of axes should zoom on that axes; up to zoom in, down to zoom out. 
- The coordinates of the mouse position should not change during the zoom: position your mouse on a notable feature of your curve (e.g. a peak), zooming in should keep that feature in place.
- Scrolling one click up should zoom in as much as scrolling one click down.
- You should be able to zoom by scrolling if any of the matplotlib tools are active (i.e. pan/zoom). Feedback on whether this is a good idea is welcome.

#### Middle-click pan
- Check that clicking and dragging using the middle button pans the figure the same way a left-click would in pan mode.
- This feature should only be active if no matplotlib tools are active.

#### Middle-click in zoom mode
- Enter zoom mode and check that a middle-click acts the same way that clicking the home button does.

Fixes #24476 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by Harrietbrown** [comment]
It might be a good idea to have clicking home be added to the states for the forward/back buttons. right now the way it works is:
1. Default (state 0)
1. Make change (state 1)
1. Click home (returns to state 0)
1. make change (state 2)
1. click back (changes to state 1)

From how I interpret it, the home button moves the state back by one then changes the layout to default. So clicking forward  after clicking home will bring you back to the state before clicking home.

**Comment by Harrietbrown** [comment]
Each wheel click counts as a new state, you may have already considered this already but could wheel clicks be grouped together like a text editor would group typing for using undo?

**Comment by Harrietbrown** [comment]
The home button works with new lines, but it does not recognize if text is added to the plot through `plt.text`

File: Framework/PythonInterface/mantid/plots/utility.py [hunk 1]
```diff
@@ -61,3 +80,52 @@ def get_autoscale_limits(ax, axis):
...
+    ax_min, ax_max = get_lims()
+    dist_to_min = coord - ax_min
+    dist_to_max = ax_max - coord
+    dist_ratio = dist_to_max/dist_to_min
+
+    new_dist_to_min = (dist_to_max - dist_to_min)/(dist_ratio - 1)/factor
+    new_dist_to_max = new_dist_to_min + (dist_to_max - dist_to_min)/factor
+
+    new_ax_min = coord - new_dist_to_min
+    new_ax_max = coord + new_dist_to_max
```
Comment (Harrietbrown) [inline Framework/PythonInterface/mantid/plots/utility.py]: The algibra can be shortened significantly here.

**Review by Harrietbrown — CHANGES_REQUESTED** [review]

**Comment by hsautessella** [comment]
I've pushed changes for the algebra and forward/backward issues.

Qt does have a [`ScrollPhase` enum](https://doc.qt.io/qt-5/qt.html#ScrollPhase-enum), on its wheelEvents. However,  every scroll event gives state 0 (`Qt::NoScrollPhase`), which means the scroll phase is not supported. It does this on both my machines, I don't know if it would be different on other machines or not. 

It is a bit annoying to have the zoom state pushed to the stack on every scroll, but I think it's better than not being able to navigate back through after scrolling, so I left it in.

As for the text labels, Matplotlib does not account for text labels when calculating limits. e.g. plotting the following, limits will only be set to centre on the lines, the text will not be in view (you'll need to pan up and to the right):
```python
import scipy.stats as stats
import math

np.random.seed(101)
num_points = 201
mu = 0
variance = 1
sigma = math.sqrt(variance)
x = np.linspace(mu - 3*sigma, mu + 3*sigma, num_points)
errs = np.random.randn(num_points)*0.01
y = stats.norm.pdf(x, mu, sigma) + errs

ws = CreateWorkspace(x, y, errs, OutputWorkspace='ws')
fig, ax = plt.subplots(subplot_kw={'projection': 'mantid'})
ax.plot(ws, specNum=1)
ax.text(3, 0.5, "Some text label", clip_on=True)
ax.legend()
plt.show()
```
Although we could have a go at implementing this if you think it would be a good idea.

**Comment by Harrietbrown** [comment]
Text labels, are not that important, they are unlikly to be there is the user hasn't specifically put it there themselves, If Matplotlib doesn't focus on them like it does lines I think it is fine for us. Good work on the home button being added to the forward/back list.
Shame about the scroll wheel but it's not essential, everything else works very well.

**Review by Harrietbrown — APPROVED** [review]
A massive improvement on the previous ability to manipulate the plot view and very intuitive.
Great work.

**Pull request merged**